src/signMsg.c

```diff
@@ -0,0 +1,389 @@
+
+void signMsg_handleInitAPDU(
+        const uint8_t* wireDataBuffer,
+        size_t wireDataSize
+)
+{
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		ctx->msgLength = parse_u4be(&view);
+		TRACE("Msg length: %d", ctx->msgLength);
+		ctx->remainingBytes = ctx->msgLength;
+
+		view_skipBytes(&view, bip44_parseFromWire(&ctx->signingPath, VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view)));
+		TRACE("Signing path:");
+		BIP44_PRINTF(&ctx->signingPath);
+		PRINTF("\n");
+
+		ctx->hashPayload = parse_bool(&view);
+		TRACE("Hash payload: %d", ctx->hashPayload);
+
+		ctx->isAscii = parse_bool(&view);
+		TRACE("Is ascii: %d", ctx->isAscii);
+
+		ctx->addressFieldType = parse_u1be(&view);
+		TRACE("Address field type: %d", ctx->addressFieldType);
+		switch (ctx->addressFieldType) {
+		case CIP8_ADDRESS_FIELD_ADDRESS:
+			view_parseAddressParams(&view, &ctx->addressParams);
+			break;
+		case CIP8_ADDRESS_FIELD_KEYHASH:
+			// no address field data to parse
+			break;
+		default:
+			THROW(ERR_INVALID_DATA);
+		}
+
+		VALIDATE(view_remainingSize(&view) == 0, ERR_INVALID_DATA);
+	}
+
+	// Check security policy
+	security_policy_t policy = policyForSignMsg(
+	                                   &ctx->signingPath,
+	                                   ctx->addressFieldType,
+	                                   &ctx->addressParams
+	                           );
+	ENSURE_NOT_DENIED(policy);
+
+	// always compute message hash
+	blake2b_224_init(&ctx->msgHashCtx);
+
+	{
+		// key is sent back at the end and possibly needed when displaying address field
+		extendedPublicKey_t extPubKey;
+		deriveExtendedPublicKey(
+		        &ctx->signingPath,
+		        &extPubKey
+		);
+		STATIC_ASSERT(SIZEOF(extPubKey.pubKey) == SIZEOF(ctx->witnessKey), "wrong witness key size");
+		memmove(ctx->witnessKey, extPubKey.pubKey, SIZEOF(extPubKey.pubKey));
+	}
+
+	// this must always be shown
+	ASSERT(policy == POLICY_PROMPT_BEFORE_RESPONSE);
+	ctx->ui_step = HANDLE_INIT_HASH_PAYLOAD;
+	signMsg_handleInit_ui_runStep();
+}
+
+static void signMsg_handleMsgChunkAPDU(const uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	{
+		ASSERT(ctx->stage == SIGN_MSG_STAGE_CHUNKS);
+		if (!ctx->hashPayload) {
+			// only a single chunk is to be received
+			ASSERT(ctx->receivedChunks == 0);
+		}
+	}
+	{
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		ctx->receivedChunks += 1;
+
+		{
+			// validates if the previous chunk has been of maximum allowed size;
+			const size_t previousSize = ctx->chunkSize;
+			if (ctx->receivedChunks == 2) {
+				if (ctx->isAscii) {
+					VALIDATE(previousSize == MAX_CIP8_MSG_FIRST_CHUNK_ASCII_SIZE, ERR_INVALID_DATA);
+				} else {
+					VALIDATE(previousSize == MAX_CIP8_MSG_FIRST_CHUNK_HEX_SIZE, ERR_INVALID_DATA);
+				}
+			} else if (ctx->receivedChunks > 2) {
+				VALIDATE(previousSize == MAX_CIP8_MSG_HIDDEN_CHUNK_SIZE, ERR_INVALID_DATA);
+			}
+		}
```

Comment:
Why do we need to perform this check for the previous chunk when the same conditions are checked below for the current chunk?